crates/anvil-polkadot/src/api_server/server.rs

```diff
@@ -1322,6 +1395,204 @@ fn transaction_matches_eth_hash(
     B256::from_slice(&tx_eth_hash) == target_eth_hash
 }
 
+/// Helper function to decode extrinsic into ETH transaction payload and signed transaction
+fn decode_eth_transaction(
+    tx_data: &Arc<polkadot_sdk::sp_runtime::OpaqueExtrinsic>,
+) -> Option<(Vec<u8>, TransactionSigned)> {
+    let encoded = tx_data.encode();
+    let ext =
+        UncheckedExtrinsic::decode_all_with_depth_limit(MAX_EXTRINSIC_DEPTH, &mut &encoded[..])
+            .ok()?;
+
+    let polkadot_sdk::sp_runtime::generic::UncheckedExtrinsic {
+        function: RuntimeCall::Revive(polkadot_sdk::pallet_revive::Call::eth_transact { payload }),
+        ..
+    } = ext.0
+    else {
+        return None;
+    };
+
+    let signed_tx = TransactionSigned::decode(&payload).ok()?;
+
+    Some((payload, signed_tx))
+}
+
+/// Fields extracted from an Ethereum transaction
+struct TransactionFields {
+    nonce: sp_core::U256,
+    to: Option<sp_core::H160>,
+    value: sp_core::U256,
+    gas: sp_core::U256,
+    gas_price: sp_core::U256,
+}
+
+/// Recover sender address from signed transaction, handling both normal and impersonated
+/// transactions
+fn recover_sender_address(_payload: &[u8], signed_tx: &TransactionSigned) -> Option<sp_core::H160> {
```

Comment:
I'd like us to avoid duplicating this logic. The `new_with_custom_address_recovery` has similar code.
Define a helper function which both code paths can use if possible.